src/moonlink_backend/src/error.rs

```diff
@@ -1,89 +1,159 @@
 use moonlink::Error as MoonlinkError;
 use moonlink_connectors::Error as MoonlinkConnectorError;
 use moonlink_connectors::PostgresSourceError;
+use moonlink_error::{ErrorStatus, ErrorStruct};
 use moonlink_metadata_store::error::Error as MoonlinkMetadataStoreError;
 use std::num::ParseIntError;
+use std::panic::Location;
 use std::result;
 use std::sync::Arc;
 use thiserror::Error;
 
 /// Custom error type for moonlink_backend
 #[derive(Clone, Debug, Error)]
 pub enum Error {
-    #[error("Parse integer error: {source}")]
-    ParseIntError { source: ParseIntError },
-    #[error("Postgres source error: {source}")]
-    PostgresSource { source: Arc<PostgresSourceError> },
-    #[error("IO error: {source}")]
-    Io { source: Arc<std::io::Error> },
-    #[error("Moonlink connector error: {source}")]
-    MoonlinkConnectorError { source: MoonlinkConnectorError },
-    #[error("Moonlink error: {source}")]
-    MoonlinkError { source: MoonlinkError },
-    #[error("Moonlink metadata error: {source}")]
-    MoonlinkMetadataStoreError { source: MoonlinkMetadataStoreError },
+    #[error("{0}")]
+    ParseIntError(ErrorStruct),
+
+    #[error("{0}")]
+    PostgresSource(ErrorStruct),
+
+    #[error("{0}")]
+    Io(ErrorStruct),
+
+    #[error("{0}")]
+    MoonlinkConnectorError(ErrorStruct),
+
+    #[error("{0}")]
+    MoonlinkError(ErrorStruct),
+
+    #[error("{0}")]
+    MoonlinkMetadataStoreError(ErrorStruct),
+
     #[error("Invalid argument: {0}")]
     InvalidArgumentError(String),
-    #[error("Watch channel receive error: {source}")]
-    TokioWatchRecvError {
-        source: tokio::sync::watch::error::RecvError,
-    },
-    #[error("JSON serialization/deserialization error: {source}")]
-    Json { source: Arc<serde_json::Error> },
+
+    #[error("{0}")]
+    TokioWatchRecvError(ErrorStruct),
+    #[error("{0}")]
+    Json(ErrorStruct),
 }
 
 pub type Result<T> = result::Result<T, Error>;
 
 impl From<PostgresSourceError> for Error {
+    #[track_caller]
     fn from(source: PostgresSourceError) -> Self {
-        Error::PostgresSource {
-            source: Arc::new(source),
-        }
+        Error::PostgresSource(ErrorStruct {
```

Comment:
I think we need to have finer error catagorization for pg error as well: https://github.com/Mooncake-Labs/moonlink/blob/743de03d5954b36e9daf9a9474a8532baee8a1eb/src/moonlink_connectors/src/pg_replicate/postgres_source.rs#L24-L43
but no need in this pr